src/protocols/routing/mod.rs

```diff
@@ -0,0 +1,270 @@
+mod forward;
+
+use serde_json::{json, Value};
+use uuid::Uuid;
+
+use crate::{
+    algorithms::AnonCryptAlg,
+    did::{DIDCommMessagingService, DIDResolver, Service, ServiceKind},
+    error::{err_msg, ErrorKind, Result, ResultContext, ResultExt},
+    message::{anoncrypt, MessagingServiceMetadata},
+    utils::did::{did_or_url, is_did},
+    Attachment, AttachmentData, Message, PackEncryptedOptions,
+};
+
+pub use self::forward::ParsedForward;
+
+pub(crate) const FORWARD_MSG_TYPE: &str = "https://didcomm.org/routing/2.0/forward";
+
+async fn find_did_comm_service<'dr>(
+    did: &str,
+    service_id: Option<&str>,
+    did_resolver: &'dr (dyn DIDResolver + 'dr),
+) -> Result<Option<Service>> {
+    let did_doc = did_resolver
+        .resolve(did)
+        .await
+        .context("Unable resolve DID")?
+        .ok_or_else(|| err_msg(ErrorKind::DIDNotResolved, "DID not found"))?;
+
+    match service_id {
+        Some(service_id) => {
+            let service: &Service = did_doc
+                .services
+                .iter()
+                .find(|&service| service.id == service_id)
+                .ok_or_else(|| {
+                    err_msg(
+                        ErrorKind::InvalidState,
+                        "Service with the specified ID not found",
+                    )
+                })?;
+
+            if let ServiceKind::DIDCommMessaging(_) = service.kind {
+                Ok(Some(service.clone()))
+            } else {
+                Err(err_msg(
+                    ErrorKind::InvalidState,
+                    "Service with the specified ID is not of DIDCommMessaging type",
+                ))
+            }
+        }
+
+        None => Ok(did_doc.services.iter().find_map(|service| {
+            if let ServiceKind::DIDCommMessaging(_) = service.kind {
+                Some(service.clone())
+            } else {
+                None
+            }
+        })),
+    }
+}
+
+fn unwrap_did_comm_service(service: &Service) -> Result<&DIDCommMessagingService> {
```

Comment:
Do we really need that method? It looks like `find_did_comm_service` can return Result<Option<DIDCommMessagingService>>.
That would match its name better (it says that it returns did_comm_service but actually returns just a service that can be non-didcomm). Also it would make error handler much easier and straightforward.

Reply:
We need `id` field, that is the reason why `find_did_comm_service` returns `Result<Option<Service>>` rather than `Result<Option<DIDCommMessagingService>>`. Because of this, we have to use the helper function `unwrap_did_comm_service` to try to extract `DIDCommMessagingService` service kind from `Service`.